### PR TITLE
Added Api calls model

### LINKED ISSUE #16: Create Api call functions model
Create and test api calls model

### PR DESCRIPTION
**Pull request by PatricioSaldivar**
Added Api Calls Model with its unit tests by mocking a http client. Also added http and mockito libraries, http is for http usage and mockito is for mocking a http client.

### CODE REVIEW

File: lib/models/api_calls.dart [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import 'package:chrome_management_app/objects/detailed_device.dart';
+
+/// The model will manage all the API calls to the Directory API.
+///
+/// See more
+/// https://developers.google.com/admin-sdk/directory/v1/guides/manage-chrome-devices.
+
+class ApiCalls {
+  ///Url to get the intial page of the list of devices
+  static final _deviceListUrl =
```
Comment (abbiePetch) [inline lib/models/api_calls.dart]: For class constant, we use DEVICE_LIST_URL

File: lib/models/api_calls.dart [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      'https://www.googleapis.com/admin/directory/v1/customer/my_customer/devices/chromeos?projection=BASIC&maxResults=15';
+
+  ///Url to get the next page of the list of devices
+  static final _deviceListWithTokenUrl =
```
Comment (abbiePetch) [inline lib/models/api_calls.dart]: same comment as above

File: lib/models/api_calls.dart [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      'https://www.googleapis.com/admin/directory/v1/customer/my_customer/devices/chromeos?projection=BASIC&maxResults=15&pageToken=';
+
+  ///Url to get detailed information of a device
+  static final _detailedDeviceUrl =
```
Comment (abbiePetch) [inline lib/models/api_calls.dart]: same comment as above

File: lib/models/api_calls.dart [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      response = await client.get(_deviceListWithTokenUrl + nextPageToken,
+          headers: {HttpHeaders.authorizationHeader: authToken});
+    }
+    // If conection is succesful return a AccountDevices object by parsing the
+    // response body
+    if (response.statusCode == 200) {
+      return AccountDevices.fromJson(json.decode(response.body));
+    } else {
+      // If that call was not successful, throw an error.
+      throw Exception('${response.statusCode} ${response.body}');
```
Comment (abbiePetch) [inline lib/models/api_calls.dart]: I assume you will handle this case in the UI (by showing the banner or msg to users that something goes wrong)

File: lib/models/api_calls.dart [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    authToken = 'Bearer $authToken';
+    final response = await client.get(_detailedDeviceUrl + deviceId,
+        headers: {HttpHeaders.authorizationHeader: authToken});
+    // If conection is succesful return a DetailedDevice object by parsing the
+    // response body
+    if (response.statusCode == 200) {
+      return DetailedDevice.fromJson(json.decode(response.body));
+    } else {
+      // If that call was not successful, throw an error.
+      throw Exception('${response.statusCode} ${response.body}');
```
Comment (abbiePetch) [inline lib/models/api_calls.dart]: same comment as above

File: lib/models/api_calls.dart [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    }
+  }
+
+  // TODO Remote commands ...
```
Comment (abbiePetch) [inline lib/models/api_calls.dart]: Please add the issue and put down the issue number or link to the issue

**Review by abbiePetch — CHANGES_REQUESTED** [review]
Good job on the code. Just a few minor things.

**Comment by pinkfish** [comment]
BTW This library: https://pub.dev/packages/built_value is useful for doing this sort of use case.  It does wrappers for dumb data objects that handles serialization and builders automatically. It is a little confusing to setup... You need to run the 'build' method to generate the classes and it is a bit finicky to get that in place to start with, however it makes creating objects way eaiser.

**Comment by PatricioSaldivar** [comment]
> BTW This library: https://pub.dev/packages/built_value is useful for doing this sort of use case. It does wrappers for dumb data objects that handles serialization and builders automatically. It is a little confusing to setup... You need to run the 'build' method to generate the classes and it is a bit finicky to get that in place to start with, however it makes creating objects way eaiser.

#34 
I made a issue to refactor the classes by using built_value library, thank you for the comment.

File: lib/models/api_calls.dart [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      response = await client.get(DEVICE_LIST_WITH_TOKEN_URL + nextPageToken,
+          headers: {HttpHeaders.authorizationHeader: authToken});
+    }
+    // If conection is succesful return a AccountDevices object by parsing the
+    // response body
+    if (response.statusCode == 200) {
+      return AccountDevices.fromJson(json.decode(response.body));
+    } else {
+      // If that call was not successful, throw an error.
+      throw Exception('${response.statusCode} ${response.body}');
```
Comment (abbiePetch) [inline lib/models/api_calls.dart]: If we are planning to show different experiences based on the status code, maybe we should throw the custom exception (or rethrow http error) with the status code, so the controller can differentiate what to do with each type of error

File: lib/models/api_calls.dart [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  /// * [authToken] is the authentication token for API to accept the requests.
+  /// * [nextPageToken]  is the next page of the devices list that should
+  /// retreive, if it is null the function will request the first page.
+  ///
+  /// This method returns a [AccountDevices] object.
+  static getDeviceList(
+      http.Client client, String authToken, String nextPageToken) async {
+    var response;
+    authToken = 'Bearer $authToken';
+    if (nextPageToken == null) {
```
Comment (Diego1149) [inline lib/models/api_calls.dart]: you can simplify this flow if you put the url inside a variable, also you might be able to use conditionals (not sure if flutter supports them):
nextPageToken==null ? url : url+'&pageToken'+token

File: lib/models/api_calls.dart [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  /// Calls the Directory API to get a list of devices with max results of 15.
+  ///
+  /// * [client] is the http client that will do the https requests.
+  /// * [authToken] is the authentication token for API to accept the requests.
+  /// * [nextPageToken]  is the next page of the devices list that should
+  /// retreive, if it is null the function will request the first page.
+  ///
+  /// This method returns a [AccountDevices] object.
+  static getDeviceList(
+      http.Client client, String authToken, String nextPageToken) async {
```
Comment (Diego1149) [inline lib/models/api_calls.dart]: You can make this class instantiable and send the client as a parameter then you don't have to make all the functions static and pass the client to each of them.

File: test/unit_tests/models/api_calls_test.dart [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+  /// Mock a authentication token
+  final authToken = '123';
+
+  /// Mock a next page token
+  final nextPageToken = '456';
+
+  /// Mock a device ID
+  final deviceId = '789';
+
+  /// Url to get the intial page of the list of devices
```
Comment (Diego1149) [inline test/unit_tests/models/api_calls_test.dart]: These 2 shouldn't be repeated, get them from the main class

File: test/unit_tests/models/api_calls_test.dart [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+  final deviceListUrl =
+      'https://www.googleapis.com/admin/directory/v1/customer/my_customer/devices/chromeos?projection=BASIC&maxResults=15';
+
+  /// Url to get the next page of the list of devices
+  final deviceListWithTokenUrl =
+      'https://www.googleapis.com/admin/directory/v1/customer/my_customer/devices/chromeos?projection=BASIC&maxResults=15&pageToken=$nextPageToken';
+
+  /// Url to get detailed information of a device
+  final detailedDeviceUrl =
```
Comment (Diego1149) [inline test/unit_tests/models/api_calls_test.dart]: same

File: lib/models/api_calls.dart (hunk 1 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Done.

File: lib/models/api_calls.dart (hunk 2 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Done.

File: lib/models/api_calls.dart (hunk 3 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Done.

File: lib/models/api_calls.dart (hunk 4 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Yes, with this method I will throw the Exception so I can catch it and manage the error.

File: lib/models/api_calls.dart (hunk 5 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Yes, with this method I will throw the Exception so I can catch it and manage the error.

File: lib/models/api_calls.dart (hunk 6 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: #38

File: lib/models/api_calls.dart (hunk 7 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Sure I'm passing the status code, so I can handle it latter depending of the code.

File: test/unit_tests/models/api_calls_test.dart (hunk 11 above)
Comment (PatricioSaldivar) [reply test/unit_tests/models/api_calls_test.dart]: Done.

File: test/unit_tests/models/api_calls_test.dart (hunk 10 above)
Comment (PatricioSaldivar) [reply test/unit_tests/models/api_calls_test.dart]: Done.

File: lib/models/api_calls.dart (hunk 9 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Done.

File: lib/models/api_calls.dart (hunk 8 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Did the ternary operation to simplify flow.

File: lib/models/api_calls.dart [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+              DEVICE_LIST_WITH_TOKEN_URL + nextPageToken,
+              headers: {HttpHeaders.authorizationHeader: authToken});
+
+      // If conection is succesful return a AccountDevices object by parsing the
+      // response body
+      if (response.statusCode == 200) {
+        return AccountDevices.fromJson(json.decode(response.body));
+      } else {
+        // If that call was not successful, throw an error.
+        throw Exception(response.statusCode);
```
Comment (abbiePetch) [inline lib/models/api_calls.dart:68]: In dart, can you do custom exception class? Please avoid the generic exception if possible

File: test/unit_tests/models/api_calls_test.dart [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    test(
+        "throws an exception if the http call for list of devices with next page token completes with an error",
+        () {
+      final client = MockClient();
+      when(client.get(ApiCalls.DEVICE_LIST_WITH_TOKEN_URL + nextPageToken,
+              headers: {HttpHeaders.authorizationHeader: 'Bearer $authToken'}))
+          .thenAnswer((_) async => http.Response('Not Found', 404));
+      final apiCall = ApiCalls(client);
+
+      expect(apiCall.getDeviceList(authToken, nextPageToken), throwsException);
```
Comment (abbiePetch) [inline test/unit_tests/models/api_calls_test.dart:138]: Is there a way to compare the error code in  the test?

**Review by abbiePetch — CHANGES_REQUESTED** [review]

File: lib/models/api_calls.dart (hunk 12 above)
Comment (pinkfish) [reply lib/models/api_calls.dart:68]: You can do custom exceptions.  There are some standard ones you can choose from too.
https://api.dart.dev/stable/2.8.4/dart-core/Error-class.html

RemoteError perhaps?

Comment (PatricioSaldivar) [reply lib/models/api_calls.dart:68]: I will create a custom exception class, called Error Handler. #35.
I will add this changes in another PR.
I'am planning to use as base Remote Error, but would like to access to a message so I can manipulate the error and just show the message as an alert and change the view if needed.

File: test/unit_tests/models/api_calls_test.dart (hunk 13 above)
Comment (PatricioSaldivar) [reply test/unit_tests/models/api_calls_test.dart:138]: With the Error Handler #35 I will, by the moment it just can catch the exception.

File: lib/models/api_calls.dart [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  static const DEVICE_LIST_URL =
+      'https://www.googleapis.com/admin/directory/v1/customer/my_customer/devices/chromeos?projection=BASIC&maxResults=15';
+
+  ///Url to get the next page of the list of devices with a next page token
+  static const DEVICE_LIST_WITH_TOKEN_URL =
+      'https://www.googleapis.com/admin/directory/v1/customer/my_customer/devices/chromeos?projection=BASIC&maxResults=15&pageToken=';
+
+  ///Url to get detailed information of a device
+  static const DETAILED_DEVICE_URL =
+      'https://www.googleapis.com/admin/directory/v1/customer/my_customer/devices/chromeos/';
```
Comment (Diego1149) [inline lib/models/api_calls.dart]: The prefix https://www.googleapis.com/admin/directory/v1 can be it's own variable

File: lib/models/api_calls.dart [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+  /// * [authToken] is the authentication token for API to accept the requests.
+  /// * [nextPageToken]  is the next page of the devices list that should
+  /// retreive, if it is null the function will request the first page.
+  ///
+  /// This method returns a [AccountDevices] object.
+  getDeviceList(String authToken, String nextPageToken) async {
+    var response;
+    authToken = 'Bearer $authToken';
+    try {
+      nextPageToken == null
```
Comment (Diego1149) [inline lib/models/api_calls.dart]: For this ones make the url conditional instead, that makes the code cleaner

**Review by Diego1149 — APPROVED** [review]

File: lib/models/api_calls.dart (hunk 14 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Done.

File: lib/models/api_calls.dart (hunk 15 above)
Comment (PatricioSaldivar) [reply lib/models/api_calls.dart]: Done, now it looks cleaner.

**Review by abbiePetch — APPROVED** [review]

**Pull request merged**